tasks/eva_3417/src/main/groovy/eva3417/RemediateLowerCaseRefAlt.groovy

```diff
@@ -0,0 +1,270 @@
+
+def options = cli.parse(args)
+if (!options) {
+    cli.usage()
+    System.exit(1)
+}
+
+// call to remediate lowercase nucleotide
+new RemediateLowerCaseNucleotide(options.prodPropertiesFile, options.hashcollisionFile).remediate()
+
+class RemediateLowerCaseNucleotide {
+    private String hashCollisionFilePath
+    private EVADatabaseEnvironment dbEnv
+    private Function<ISubmittedVariant, String> hashingFunction
+
+    static def logger = LoggerFactory.getLogger(Application.class)
+
+    RemediateLowerCaseNucleotide(prodPropertiesFile, hashcollisionFile) {
+        this.dbEnv = createFromSpringContext(prodPropertiesFile, Application.class)
+        this.hashCollisionFilePath = hashcollisionFile
+        this.hashingFunction = new SubmittedVariantSummaryFunction().andThen(new SHA1HashingFunction())
+    }
+
+    List<SubmittedVariantEntity> getImpactedSVEList(List<SubmittedVariantEntity> sveList) {
+        List<SubmittedVariantEntity> impactedSVEList = new ArrayList<>()
+        for (SubmittedVariantEntity sve : sveList) {
+            String ref = sve.getReferenceAllele()
+            String alt = sve.getAlternateAllele()
+            if (ref != ref.toUpperCase() || alt != alt.toUpperCase()) {
+                impactedSVEList.add(sve)
+            }
+        }
+
+        return impactedSVEList
+    }
+
+
+    String getSVENewHash(SubmittedVariantEntity sve) {
+        SubmittedVariant temp = sve.getModel()
+        temp.setReferenceAllele(sve.getReferenceAllele().toUpperCase())
+        temp.setAlternateAllele(sve.getAlternateAllele().toUpperCase())
+        return hashingFunction.apply(temp)
+    }
+
+    List<String> getExistingHashList(Collection<String> newHashList) {
+        List<SubmittedVariantEntity> existingHashSVEList = dbEnv.mongoTemplate
+                .find(query(where("_id").in(newHashList)), sveClass)
+
+        return existingHashSVEList.stream()
+                .map(sve -> sve.getHashedMessage())
+                .collect(Collectors.toList())
+    }
+
+    void insertSVEWithNewHash(List<SubmittedVariantEntity> sveList, Map<String, String> mapOldHashNewHash) {
+        List<SubmittedVariantEntity> sveWithNewHashList = new ArrayList<>()
+        for (SubmittedVariantEntity sve : sveList) {
+            // create a new SVE with updated ref, alt and hash
+            SubmittedVariant sveModel = sve.getModel()
+            sveModel.setReferenceAllele(sve.getReferenceAllele().toUpperCase())
+            sveModel.setAlternateAllele(sve.getAlternateAllele().toUpperCase())
+            SubmittedVariantEntity submittedVariantEntity = new SubmittedVariantEntity(sve.getAccession(),
+                    mapOldHashNewHash.get(sve.getHashedMessage()), sveModel, sve.getVersion())
+
+            sveWithNewHashList.add(submittedVariantEntity)
+        }
+
+        // insert updated SVEs into db
+        dbEnv.mongoTemplate.insert(sveWithNewHashList, sveClass)
```

Comment:
Did you try "mongoTemplate.save" here instead? it can serve as an "upsert" and update documents when they are already there but insert them when they are not...

Reply:
Also, when there is actual hash collision involved and a merge needs to happen, you can use [an operation to record that](https://github.com/EBIvariation/eva-tasks/blob/278c5387d15e32271af451db01ef3c100a00fb27/tasks/eva_3399/src/main/groovy/eva3399/DeprecateOrphanedRS.groovy#L57)

Reply:
added